plugin/core/collections.py

```diff
@@ -135,7 +135,17 @@ def update(self, d: Dict[str, Any]) -> None:
         """
         Overwrite and/or add new key-value pairs to the collection.
 
+        :param      d:    The overriding dictionary. Keys must be in the new-style dotted format.
+        """
+        for key, value in d.items():
+            self.set(key, value)
+
+    def update_recursive(self, d: Dict[str, Any]) -> None:
+        """
+        Overwrite and/or add new key-value pairs to the collection.
+
         :param      d:    The overriding dictionary. Can contain nested dictionaries.
+                          The key is dynamically built up from pieces.
```

Comment:
When we have default settings (in this example LSP-eslint):

```js
{
	"settings": {
		"codeAction.disableRuleComment": {
			"enable": true,
			"location": "separateLine"
		},
}
```

and we want to override just the `enable` setting with:

```js
{
    "settings": {
        "codeAction.disableRuleComment": {
            "enable": true
        },
}
```

then the `settings.codeAction.disableRuleComment.location` will be gone in the resulting object.

I'm guessing that it's not how it's supposed to work? In VSCode the settings are merged correctly.

Reply:
Overriding with `"codeAction.disableRuleComment.enable": true,` works but it wouldn't match the schema so would trigger a warning diagnostic.

Reply:
So I looked into this and we can't have both your above described behavior _and_ fix the `http://foo.com -> http://foo  com` bug, unless we start requiring a schema and parsing that schema.

Reply:
I'm still quite confused about how VSCode handles it. I don't think it actually uses schema for that. At least haven't found any evidence of that when checking out its code before.

Reply:
It looks to me like VSCode just doesn't do expansion recursively.

I've tested a bit on the `yaml` server in VSCode. Defined settings like:

```js
  "yaml.schemas": {
    "test.me": "foo"
  },
  "yaml.schemas.foo": {
    "test.me": "foo"
  },
  "yaml.foo": {
    "test.me": "abc"
  },
```

Where only the first setting matches the schema. This resulted in server receiving these settings in `workspace/configuration` request:

```js
        "schemas": {
            "test.me": "foo",
            "foo": {
                "test.me": "foo"
            }
        },
        "foo": {
            "test.me": "abc"
        }
```

So schema doesn't seem to matter here. What it seems to do is that it only expands the keys at the root of the `settings` object.

Reply:
I'm thinking that we could possibly **only** support dotted keys at the root level of the `settings` object. Possibly not even in `initializationSettings`. Though ideally one would go through all LSP(-*) configuration files to see if that is likely to break anything.